### DISCUSSION TITLE
No applicable cell found (string array) for jsonforms-property-view

### DISCUSSION THREAD

**Discussion by hsd-dev** [eclipse-emfcloud/emfcloud#178] [All other] [answered]
I am using `jsonforms-property-view` [0.8.0-next.78f8ae13](https://www.npmjs.com/package/@eclipse-emfcloud/jsonforms-property-view/v/0.8.0-next.78f8ae13) version (from [here](https://github.com/eclipse-emfcloud/emfcloud/discussions/126#discussioncomment-2318422)). For a simple model
```
{"name":"NewLink0","id":"95e0cf3b-5e34-4f5d-9a4d-b6a21c6b9af0"}
```
the property view is rendered as expected

![Screenshot from 2022-09-20 23-49-42](https://user-images.githubusercontent.com/31062848/191371121-fb8eb816-04cb-4b1f-aed4-eee11eb04bf7.png)

But when a model has string elements within an array
```
{"name":"NewLink0","id":"e970b3cb-0c18-485a-a187-20475380815c","xyz":{"abc":"0 0 0","pqr":"0 0 0"}}
```
I get an error
```
root WARN No applicable cell found. { type: 'Control', scope: '#/properties/xyz' } {
  type: 'object',
  properties: { abc: { type: 'string' }, pqr: { type: 'string' } },
  additionalProperties: true,
  required: [ 'abc', 'pqr' ]
}
```
and property view is empty. I found a possibly related issue: https://github.com/eclipsesource/jsonforms/issues/1350

I have tried the later versions:
[0.8.0-next.bc15ec6.30](https://www.npmjs.com/package/@eclipse-emfcloud/jsonforms-property-view/v/0.8.0-next.bc15ec6.30)
[0.8.0-next.85882ad2](https://www.npmjs.com/package/@eclipse-emfcloud/jsonforms-property-view/v/0.8.0-next.85882ad2)
[0.8.0-next.8ca4b926](https://www.npmjs.com/package/@eclipse-emfcloud/jsonforms-property-view/v/0.8.0-next.8ca4b926)
[0.8.0-next.91b431f6](https://www.npmjs.com/package/@eclipse-emfcloud/jsonforms-property-view/v/0.8.0-next.91b431f6)
[0.8.0-next.30ecf115](https://www.npmjs.com/package/@eclipse-emfcloud/jsonforms-property-view/v/0.8.0-next.30ecf115)

either with the same error, or I don't see the jsonforms at all, even for a simple model.

**Comment by sdirix** [comment, marked as answer]
Hi @ipa-hsd,

I don't really see an array of strings in the example data. In your case the property `xyz` seems to be a nested object. The property view is currently meant for flat data and does not supported nested objects. This might change in the future.

Besides restructuring your data there are two ways to handle this:
 - Specify a UI Schema for your object which is able to handle the nested objects, e.g.

```
{
  "type": "VerticalLayout",
  "elements": [
    {
      "type": "Control",
      "scope": "#/properties/name"
    },
    {
      "type": "Control",
      "scope": "#/properties/id"
    },
    {
      "type": "Control",
      "scope": "#/properties/xyz/properties/abc"
    },
    {
      "type": "Control",
      "scope": "#/properties/xyz/properties/pqr"
    }
  ]
}
```
- Or you extend the React Vanilla JSON Forms renderer set which is used in the properties view. At the moment it's not able to handle controls which point to objects which is why you get the `no applicable cell` found issue. However it's rather simple to implement, as you basically just need to re-delegate to JSON Forms for all nested properties. If you need this feel free to check the [JSON Forms forum](https://jsonforms.discourse.group/) in which we can help you in creating the custom renderer or even create a contribution.

**Reply by hsd-dev** [reply]
Thanks! So for the first option, I need to implement [this function](https://github.com/eclipse-emfcloud/ecore-glsp/blob/3f0d45204371d4aaef0f43bd8cdcf2c6b2bda082/client/packages/ecore-property-view/src/browser/property-data-service.ts#L86) in my `property-data-service`?
```
    async getUiSchema(selection: any, properties?: any): Promise<UISchemaElement | undefined> {
        if (properties && properties.eClass) {
            const eClassName = this.getSelectionData(selection).eClass || new URI(properties.eClass).fragment.substring(2);
            return this.modelServerClient.getUiSchema(eClassName.toLowerCase()).then((response: any) => response.body as UISchemaElement);
        }
        return undefined;
    }
```

**Reply by sdirix** [reply]
Yes you can either generate or retrieve your UI schema directly here or let it deliver via the model server (as you can see in the code). When initially configuring the model server it allows for a directory with UI Schemas to be handed over which it will then deliver. The filename of the UI Schema has to correspond to the eClass it's meant for.

**Reply by hsd-dev** [reply]
Currently I am not using model server (would be great to have a project template to set it up). I have implemented my own custom action handlers.
> The filename of the UI Schema has to correspond to the eClass it's meant for.

Noted!

**Reply by hsd-dev** [reply]
works now!
